File: lib/replay.mjs
import path from 'node:path';
import { writeFile } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { sha256Hex, sha256File, applySubstitutions, runCommand } from './util.mjs';
import { appendRecord, readChain, ARTIFACTS_DIR } from './chain.mjs';

export async function runReplay(cwd, captureId) {
  const records = await readChain(cwd);
  const capture = records.find((r) => r.id === captureId && r.type === 'capture');
  if (!capture) throw new Error(`capture not found: ${captureId}`);
  const captured = capture.payload;

  const inputChecks = [];
  for (const file of captured.inputFiles) {
    let actual = null;
    try {
      actual = await sha256File(path.join(cwd, file.path));
    } catch {
      // missing file
    }
    inputChecks.push({ path: file.path, expected: file.sha256, actual, match: actual === file.sha256 });
  }
  const inputsChanged = inputChecks.some((check) => !check.match);

  // Recorded output files are evidence artifacts: check them BEFORE re-running
  // the command, which would otherwise rewrite the very state under test.
  const outputFields = [];
  for (const file of captured.outputFiles) {
    let actual = null;
    try {
      actual = await sha256File(path.join(cwd, file.path));
    } catch {
      // missing file
    }
    outputFields.push({ field: `out:${file.path}`, expected: file.sha256, actual, match: file.sha256 === actual });
  }

  if (!existsSync(captured.cwd)) {
    throw new Error(`recorded working directory no longer exists: ${captured.cwd}`);
  }

  const { stdout, stderr, exitCode } = await runCommand(captured.command, captured.cwd);
  const stdoutFinal = applySubstitutions(applySubstitutions(stdout, captured.redactions), captured.normalizers);
  const stderrFinal = applySubstitutions(applySubstitutions(stderr, captured.redactions), captured.normalizers);

  const stdoutHash = sha256Hex(stdoutFinal);
  const stderrHash = sha256Hex(stderrFinal);
  const fields = [
    { field: 'exitCode', expected: captured.exitCode, actual: exitCode, match: captured.exitCode === exitCode },
    { field: 'stdout', expected: captured.stdoutHash, actual: stdoutHash, match: captured.stdoutHash === stdoutHash },
    { field: 'stderr', expected: captured.stderrHash, actual: stderrHash, match: captured.stderrHash === stderrHash },
    ...outputFields,
  ];

  const allMatch = fields.every((field) => field.match);
  const verdict = inputsChanged ? 'INVALID_INPUTS' : allMatch ? 'MATCH' : 'DRIFT';

  const record = await appendRecord(cwd, 'replay', {
    ref: captureId,
    verdict,
    inputsChanged,
    fields,
    runtime: { platform: process.platform, node: process.versions.node },
  });

  await writeFile(path.join(cwd, ARTIFACTS_DIR, `${record.id}.stdout.txt`), stdoutFinal, 'utf8');
  await writeFile(path.join(cwd, ARTIFACTS_DIR, `${record.id}.stderr.txt`), stderrFinal, 'utf8');

  return { record, verdict, fields, inputChecks };
}

File: cpp/replay.cpp
// replay.cpp — see replay.h.
#include "replay.h"

#include "chain.h"
#include "jsjson.h"
#include "jsre.h"
#include "sha256.h"
#include "util.h"

#include <stdexcept>

namespace pocev {

namespace {

std::string payloadString(const Json* payload, const char* key, const std::string& fallback = "") {
    if (!payload) return fallback;
    const Json* v = payload->get(key);
    return v && v->isStr() ? v->str : fallback;
}

}  // namespace

ReplayResult runReplay(const std::string& cwd, const std::string& captureId) {
    std::vector<Json> records = readChain(cwd);
    const Json* capture = nullptr;
    for (const auto& r : records) {
        const Json* id = r.get("id");
        const Json* type = r.get("type");
        if (id && id->isStr() && id->str == captureId && type && type->isStr() &&
            type->str == "capture") {
            capture = &r;
            break;
        }
    }
    if (!capture) throw std::runtime_error("capture not found: " + captureId);
    const Json* captured = capture->get("payload");

    Json inputChecks = Json::array();
    const Json* inputFiles = captured->get("inputFiles");
    bool inputsChanged = false;
    if (inputFiles && inputFiles->isArr()) {
        for (const auto& file : inputFiles->arr) {
            const Json* path = file.get("path");
            const Json* expected = file.get("sha256");
            std::string actual;
            bool present = false;
            if (path && path->isStr()) {
                try {
                    actual = sha256File(pathJoin(cwd, path->str));
                    present = true;
                } catch (const std::runtime_error&) {
                }
            }
            bool match = present && expected && expected->isStr() && expected->str == actual;
            if (!match) inputsChanged = true;
            Json check = Json::object();
            check.set("path", path ? *path : Json::null());
            check.set("expected", expected ? *expected : Json::null());
            check.set("actual", present ? Json::string(actual) : Json::null());
            check.set("match", Json::boolean(match));
            inputChecks.push(check);
        }
    }

    Json outputFields = Json::array();
    const Json* outputFiles = captured->get("outputFiles");
    if (outputFiles && outputFiles->isArr()) {
        for (const auto& file : outputFiles->arr) {
            const Json* path = file.get("path");
            const Json* expected = file.get("sha256");
            std::string actual;
            bool present = false;
            if (path && path->isStr()) {
                try {
                    actual = sha256File(pathJoin(cwd, path->str));
                    present = true;
                } catch (const std::runtime_error&) {
                }
            }
            bool match = present && expected && expected->isStr() && expected->str == actual;
            Json f = Json::object();
            f.set("field", Json::string("out:" + (path && path->isStr() ? path->str : "")));
            f.set("expected", expected ? *expected : Json::null());
            f.set("actual", present ? Json::string(actual) : Json::null());
            f.set("match", Json::boolean(match));
            outputFields.push(f);
        }
    }

    std::string recordedCwd = payloadString(captured, "cwd");
    if (!pathExists(recordedCwd)) {
        throw std::runtime_error("recorded working directory no longer exists: " + recordedCwd);
    }

    std::vector<std::string> command;
    const Json* commandArr = captured->get("command");
    if (commandArr && commandArr->isArr()) {
        for (const auto& c : commandArr->arr) {
            if (c.isStr()) command.push_back(c.str);
        }
    }
    CommandResult run = runCommand(command, recordedCwd);
    std::vector<Substitution> redactions, normalizers;
    auto readSubs = [](const Json* list, std::vector<Substitution>& out) {
        if (!list || !list->isArr()) return;
        for (const auto& s : list->arr) {
            Substitution sub;
            const Json* p = s.get("pattern");
            const Json* f = s.get("flags");
            const Json* r = s.get("replacement");
            sub.pattern = p && p->isStr() ? p->str : "";
            sub.flags = f && f->isStr() ? f->str : "";
            sub.replacement = r && r->isStr() ? r->str : "";
            out.push_back(sub);
        }
    };
    readSubs(captured->get("redactions"), redactions);
    readSubs(captured->get("normalizers"), normalizers);
    std::string stdoutFinal = applySubstitutions(applySubstitutions(run.out, redactions),
                                                normalizers);
    std::string stderrFinal = applySubstitutions(applySubstitutions(run.err, redactions),
                                                normalizers);

    std::string stdoutHash = sha256Hex(stdoutFinal);
    std::string stderrHash = sha256Hex(stderrFinal);

    Json fields = Json::array();
    auto addField = [&](const std::string& name, const Json* expected, Json actual, bool match) {
        Json f = Json::object();
        f.set("field", Json::string(name));
        f.set("expected", expected ? *expected : Json::null());
        f.set("actual", std::move(actual));
        f.set("match", Json::boolean(match));
        fields.push(f);
    };
    const Json* exitExpected = captured->get("exitCode");
    Json exitActual = run.signaled ? Json::null() : Json::number(static_cast<double>(run.code));
    // captured.exitCode === exitCode, strict: null === null matches too.
    bool exitMatch = false;
    if (exitExpected && exitExpected->isNum() && !run.signaled) {
        exitMatch = exitExpected->num == static_cast<double>(run.code);
    } else if (exitExpected && exitExpected->isNull() && run.signaled) {
        exitMatch = true;
    }
    addField("exitCode", exitExpected, exitActual, exitMatch);
    const Json* stdoutExpected = captured->get("stdoutHash");
    addField("stdout", stdoutExpected, Json::string(stdoutHash),
             stdoutExpected && stdoutExpected->isStr() && stdoutExpected->str == stdoutHash);
    const Json* stderrExpected = captured->get("stderrHash");
    addField("stderr", stderrExpected, Json::string(stderrHash),
             stderrExpected && stderrExpected->isStr() && stderrExpected->str == stderrHash);
    for (const auto& f : outputFields.arr) fields.push(f);

    bool allMatch = true;
    for (const auto& f : fields.arr) {
        const Json* m = f.get("match");
        if (!m || !m->isBool() || !m->b) {
            allMatch = false;
            break;
        }
    }
    std::string verdict = inputsChanged ? "INVALID_INPUTS" : (allMatch ? "MATCH" : "DRIFT");

    Json payload = Json::object();
    payload.set("ref", Json::string(captureId));
    payload.set("verdict", Json::string(verdict));
    payload.set("inputsChanged", Json::boolean(inputsChanged));
    payload.set("fields", fields);
    Json runtime = Json::object();
#ifdef _WIN32
    runtime.set("platform", Json::string("win32"));
#elif defined(__APPLE__)
    runtime.set("platform", Json::string("darwin"));
#else
    runtime.set("platform", Json::string("linux"));
#endif
    runtime.set("node", Json::string("cpp"));
    payload.set("runtime", runtime);

    Json record = appendRecord(cwd, "replay", payload);

    writeFileBytes(
        pathJoin(pathJoin(cwd, ARTIFACTS_DIR), record.get("id")->str + ".stdout.txt"),
        stdoutFinal);
    writeFileBytes(
        pathJoin(pathJoin(cwd, ARTIFACTS_DIR), record.get("id")->str + ".stderr.txt"),
        stderrFinal);

    ReplayResult out;
    out.record = record;
    out.verdict = verdict;
    out.fields = fields;
    out.inputChecks = inputChecks;
    return out;
}

}  // namespace pocev
